### PR TITLE
[e2e] Test CSC with non-existant targetNamespace

### PR DESCRIPTION
**Pull request by awgreene**
Implement a test that:
* Creates a CatalogSourceConfig with non-existent targetNamespace
* Ensures that the status is not succeeded
* Creates the targetNamespace
* Ensures that the status is succeeded

Create a testutils file that contains methods used by various tests

### CODE REVIEW

File: test/e2e/marketplace_test.go [hunk 1]
```diff
@@ -123,9 +119,6 @@ func defaultCreateTest(t *testing.T, f *test.Framework, ctx *test.TestCtx) error
 			Timeout:       cleanupTimeout,
 			RetryInterval: cleanupRetryInterval,
 		})
```
Comment (kevinrizza) [inline test/e2e/marketplace_test.go]: why are we removing this error check?

File: test/e2e/testutils.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+			Namespace: namespace,
+		},
+		Spec: operator.CatalogSourceConfigSpec{
+			TargetNamespace: targetNamespace,
+			Packages:        packages,
+		}}
+}
+
+// createOperatorSource returns a OperatorSource based on the provided parameters
+func createOperatorSource(operatorSourceName string, namespace string, labels map[string]string, registryNamespace string) *operator.OperatorSource {
```
Comment (kevinrizza) [inline test/e2e/testutils.go]: What is the value of this method? What if we wanted to create an operator source with a different name? Or an invalid namespace? IMO if there was logic in this method it would be one thing, but we could just use this syntax inline in the method and have more control without much more verbosity.

File: test/e2e/testutils.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	cleanupRetryInterval = time.Second * 1
+	cleanupTimeout       = time.Second * 30
+)
+
+// createCatalogSourceConfig returns a CatalogSourceConfig based on the provided parameters
+func createCatalogSourceConfig(catalogSourceConfigName string, namespace string, targetNamespace string, packages string) *operator.CatalogSourceConfig {
```
Comment (kevinrizza) [inline test/e2e/testutils.go]: Same comment as the one about createOperatorSource

File: test/e2e/testutils.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+		})
+}
+
+// pollUntilTrue will check if a function returns a true value
+// every pollInterval duration until the timeout duration has
+// passed
+//
+// pollUntilTrue function will run over the timeout if the
+// provided function hangs
+func pollUntilTrue(pollInterval time.Duration, timeout time.Duration, f func() bool) bool {
```
Comment (kevinrizza) [inline test/e2e/testutils.go]: Don't we already have a very similar method to this in helpers.go?

File: test/e2e/testutils.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
+package e2e
```
Comment (kevinrizza) [inline test/e2e/testutils.go]: Why don't we use the `helpers.go` file for this? Why do we need a second utilities/helpers file?

File: test/e2e/marketplace_test.go [hunk 6]
```diff
@@ -74,7 +71,6 @@ func MarketplaceCluster(t *testing.T) {
 
 	// get global framework variables
```
Comment (kevinrizza) [inline test/e2e/marketplace_test.go]: If we are going to have multiple tests and multiple files for the tests, wouldn't it be better to move this basic workflow test into a separate file as well?

File: test/e2e/csc_tests.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	registryNamespace := "marketplace_e2e"
+
+	// Get test namespace
+	namespace, err := ctx.GetNamespace()
+	if err != nil {
+		return fmt.Errorf("Could not get namespace: %v", err)
+	}
+
+	// Create the operatorsource to download the manifests
+	testOperatorSource := createOperatorSource(testResourceName, namespace, nil, registryNamespace)
```
Comment (kevinrizza) [inline test/e2e/csc_tests.go]: For CatalogSourceConfig tests, if we are running these in serial, would it be better if we did some upfront setup like create one operatorsource beforehand, and then use the same one for all the tests? Rather than create all these resources each time for each test?

File: test/e2e/csc_tests.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+	// Check that we created the catalogsourceconfig with an invalid targetNamespace
+	err = WaitForResult(t, f, resultCatalogSourceConfig, namespace, testResourceName)
+	if err != nil {
+		return err
+	}
+
+	// Check if the catalogsourceconfig phase and message are the expected values
+	expectedPhase := "Configuring"
+	expectedMessage := fmt.Sprintf("namespaces \"%s\" not found", targetNamespace)
+	result := pollUntilTrue(5*time.Second, 1*time.Minute, func() bool {
```
Comment (kevinrizza) [inline test/e2e/csc_tests.go]: Don't we have these timeouts hardcoded elsewhere?

File: test/e2e/testutils.go (hunk 4 above)
Comment (awgreene) [reply test/e2e/testutils.go]: In the helpers.go there exists a tool to check that a runtime object exists. There is not a way to check that a runtime Object has reached a certain state.

File: test/e2e/testutils.go (hunk 5 above)
Comment (awgreene) [reply test/e2e/testutils.go]: Good point - I'll move any methods into the existing `helpers.go` file

File: test/e2e/marketplace_test.go (hunk 6 above)
Comment (awgreene) [reply test/e2e/marketplace_test.go]: I am planning to create another PR that would refactor the basic workflow test into it's own file. I thought it made sense to limit the scope of this PR to introducing the csc test.

File: test/e2e/marketplace_test.go (hunk 1 above)
Comment (awgreene) [reply test/e2e/marketplace_test.go]: This should not have been removed - my mistake.

File: test/e2e/csc_tests.go (hunk 7 above)
Comment (awgreene) [reply test/e2e/csc_tests.go]: This is certainly something I had considered. I plan to do a follow up PR that refactors existing tests which would include this work.

File: test/e2e/csc_tests.go (hunk 8 above)
Comment (awgreene) [reply test/e2e/csc_tests.go]: There are constants included in the testutils package that define the interval at which we attempt to create a runtime object and the duration we should attempt to do so for. I could update these calls to use those values but the timeout value would decrease which means the tests may fail more. At the very least - the values should be updated to be constants.

File: test/e2e/testutils.go (hunk 2 above)
Comment (awgreene) [reply test/e2e/testutils.go]: I see your point. I will remove the methods in the interest of having more control over the objects we create for the tests.

File: test/e2e/testutils.go (hunk 3 above)
Comment (awgreene) [reply test/e2e/testutils.go]: Please see comment on createOperatorSource.

File: test/e2e/testutils.go (hunk 5 above)
Comment (awgreene) [reply test/e2e/testutils.go]: Moved.

File: test/e2e/csc_tests.go (hunk 8 above)
Comment (aravindhp) [reply test/e2e/csc_tests.go]: Yes, please add constants with comments for these.

File: test/e2e/csc_tests.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+			Kind: operator.OperatorSourceKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-operators",
+			Namespace: namespace,
+		},
+		Spec: operator.OperatorSourceSpec{
+			Type:              "appregistry",
+			Endpoint:          "https://quay.io/cnr",
+			RegistryNamespace: "marketplace_e2e",
```
Comment (aravindhp) [inline test/e2e/csc_tests.go]: Given we are going to constantly add and remove this `OperatorSource`, please reduce the number of operators in the namespace.

File: test/e2e/csc_tests.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+	// Check that we created the catalogsourceconfig
+	resultCatalogSourceConfig := &operator.CatalogSourceConfig{}
+	err = WaitForResult(t, f, resultCatalogSourceConfig, namespace, testResourceName)
+	if err != nil {
+		return err
+	}
+
+	// Delete the existing catalogsourceconfig
+	err = deleteRuntimeObject(f, resultCatalogSourceConfig, testResourceName, namespace)
```
Comment (aravindhp) [inline test/e2e/csc_tests.go]: Why are you deleting the datastore `CatalogSourceConfig`?

File: test/e2e/csc_tests.go [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+		}
+		return false
+	})
+
+	if !result {
+		return fmt.Errorf("CatalogSourceConfig never reached expected phase/message, expected %v/%v got %v/%v", expectedPhase, expectedMessage, resultCatalogSourceConfig.Status.CurrentPhase.Name, resultCatalogSourceConfig.Status.CurrentPhase.Message)
+	}
+
+	// Create the nonexisting targetNamespace
+	err = createNamespace(f, ctx, targetNamespace)
```
Comment (aravindhp) [inline test/e2e/csc_tests.go]: Does the `e2e` launched using the SDK use a super user account? Because the operator does not have `Roles` that allow it to interact with namespaces. This seems to imply that there is no way for us to verify if we are adding the correct `Roles` for the operator.

File: test/e2e/helpers.go [hunk 12]
```diff
@@ -71,3 +76,80 @@ func WaitForSuccessfulDeployment(t *testing.T, f *test.Framework, deployment app
...
+	// Delete the runtime object
+	err := f.Client.Delete(
+		context.TODO(),
+		obj)
+	if err != nil {
+		return fmt.Errorf("Error deleting obj: %v", err)
+	}
+
+	// Confirm that the runtime object was deleted
+	namespacedName := types.NamespacedName{Name: name, Namespace: namespace}
```
Comment (aravindhp) [inline test/e2e/helpers.go]: Why is this function also confirming it was deleted while `createRuntimeObject` does only the creation? What if I have a use where I just want to delete an object and not care if it actually got deleted?

File: test/e2e/helpers.go [hunk 13]
```diff
@@ -71,3 +76,80 @@ func WaitForSuccessfulDeployment(t *testing.T, f *test.Framework, deployment app
...
+		return nil
+	}
+	return fmt.Errorf("Runtime object %v was never deleted", obj)
+}
+
+// createNamespace creates a namespace
+func createNamespace(f *test.Framework, ctx *test.TestCtx, namespace string) error {
+	obj := &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}}
+	// Create the operatorsource to download the manifests
+	return f.Client.Create(
```
Comment (aravindhp) [inline test/e2e/helpers.go]: Please reuse `createRuntimeObject()` here?

File: test/e2e/helpers.go [hunk 14]
```diff
@@ -71,3 +76,80 @@ func WaitForSuccessfulDeployment(t *testing.T, f *test.Framework, deployment app
...
+		})
+}
+
+// pollUntilTrue will check if a function returns a true value
+// every pollInterval duration until the timeout duration has
+// passed
+//
+// pollUntilTrue function will run over the timeout if the
+// provided function hangs
+func pollUntilTrue(pollInterval time.Duration, timeout time.Duration, f func() bool) bool {
```
Comment (aravindhp) [inline test/e2e/helpers.go]: Are you sure a similar helper doesn't exist in the SDK or k8s libraries?

File: test/e2e/marketplace_test.go [hunk 15]
```diff
@@ -32,14 +28,14 @@ func TestMarketplace(t *testing.T) {
 		TypeMeta: metav1.TypeMeta{
 			Kind: marketplace.OperatorSourceKind,
 			APIVersion: fmt.Sprintf("%s/%s",
-			marketplace.SchemeGroupVersion.Group, marketplace.SchemeGroupVersion.Version),
+				marketplace.SchemeGroupVersion.Group, marketplace.SchemeGroupVersion.Version),
```
Comment (aravindhp) [inline test/e2e/marketplace_test.go]: Formatting fixes should be a separate PR

File: test/e2e/marketplace_test.go [hunk 16]
```diff
@@ -32,14 +28,14 @@ func TestMarketplace(t *testing.T) {
...
 		},
 	}
 	catalogsourceconfig := &marketplace.CatalogSourceConfig{
 		TypeMeta: metav1.TypeMeta{
 			Kind: marketplace.CatalogSourceConfigKind,
 			APIVersion: fmt.Sprintf("%s/%s",
-			marketplace.SchemeGroupVersion.Group, marketplace.SchemeGroupVersion.Version),
+				marketplace.SchemeGroupVersion.Group, marketplace.SchemeGroupVersion.Version),
```
Comment (aravindhp) [inline test/e2e/marketplace_test.go]: Formatting fixes should be a separate PR

File: test/e2e/marketplace_test.go [hunk 17]
```diff
@@ -50,6 +46,7 @@ func TestMarketplace(t *testing.T) {
 	if err != nil {
 		t.Fatalf("failed to add catalogsourceconfig custom resource scheme to framework: %v", err)
 	}
+
```
Comment (aravindhp) [inline test/e2e/marketplace_test.go]: Why? Please don't add and remove empty lines because you prefer it that way. We can come up with a guideline for such things and fix all of this in a separate PR.

File: test/e2e/csc_tests.go (hunk 11 above)
Comment (kevinrizza) [reply test/e2e/csc_tests.go]: @aravindhp There are two competing ideas here. the test client has similar super user rules that apply to it, similar to what `operator-sdk up` has (or akin to logging into oc as a super user), which is what is being used to create this namespace. That is separate from the operator itself having those roles, which are being tested here -- the operator is deployed using the deployment in the manifest/ folder when these tests are run in the e2e cluster.

Comment (aravindhp) [reply test/e2e/csc_tests.go]: So your take is that the `ClusterOperator` will be marked as failing if the roles are incorrect? We should then update our deployment check to include that and not run any other tests if that fails.

Comment (kevinrizza) [reply test/e2e/csc_tests.go]: It should, but I think that's outside the scope of this test -- ClusterOperator having a failing status is a catch all, so we could lose granularity on why the error is happening if we just make that blanket assumption. Better for us to ignore ClusterOperator status, or at least just mark a failure and continue the tests. What if we want to run a test to purposely make ClusterOperator set a failed status? Or run a test to see a fail case that we would ignore because we are looking at ClusterOperator?

Comment (aravindhp) [reply test/e2e/csc_tests.go]: Good point. I did not mean to have this be done as part of this PR. We should have a follow up PR for this. My desire is to have a test that will tell us if we messed up the `Roles` in the manifest folder for the operator.

Comment (kevinrizza) [reply test/e2e/csc_tests.go]: +1 I like the idea of explicit tests that check and ensure the state of OperatorStatus.

File: test/e2e/csc_tests.go (hunk 9 above)
Comment (awgreene) [reply test/e2e/csc_tests.go]: Rather than creating an operator source for each test, would it be better to use a single operator source for all tests that rely on it? I had spoken to Kevin about doing so in a future refactor PR: https://github.com/operator-framework/operator-marketplace/pull/145#discussion_r269083130

Comment (aravindhp) [reply test/e2e/csc_tests.go]: +1

Comment (awgreene) [reply test/e2e/csc_tests.go]: Great - I will create a JIRA to refactor the code.

File: test/e2e/csc_tests.go (hunk 10 above)
Comment (awgreene) [reply test/e2e/csc_tests.go]: After looking into this I realized that I shouldn't delete this CSC. I can just create the CSC that points to an invalid `targetNamespace`.

This also means that the `deleteRuntimeObject` method in the `helper.go` and so I will remove it.

File: test/e2e/helpers.go (hunk 14 above)
Comment (awgreene) [reply test/e2e/helpers.go]: The SDK only provides the controller-runtime's client and the `WaitForDeployment` and `WaitForOperatorDeployment` functions in the test framework. I will look deeper into the k8s library and update this accordingly.

File: test/e2e/marketplace_test.go (hunk 16 above)
Comment (awgreene) [reply test/e2e/marketplace_test.go]: This was a mistake introduced by the go format when saving the file, I will revert the change.

File: test/e2e/marketplace_test.go (hunk 15 above)
Comment (awgreene) [reply test/e2e/marketplace_test.go]: This was a mistake introduced by the go format when saving the file, I will revert the change.

File: test/e2e/marketplace_test.go (hunk 17 above)
Comment (awgreene) [reply test/e2e/marketplace_test.go]: Addressed in the latest commit.

File: test/e2e/helpers.go (hunk 12 above)
Comment (awgreene) [reply test/e2e/helpers.go]: After reviewing a comment you made earlier[1] this function was removed given that the test is not required to delete a csc.

File: test/e2e/helpers.go (hunk 13 above)
Comment (awgreene) [reply test/e2e/helpers.go]: Addressed in the latest commit.

File: test/e2e/csc_tests.go [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+
+	operator "github.com/operator-framework/operator-marketplace/pkg/apis/operators/v1"
+	"github.com/operator-framework/operator-sdk/pkg/test"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// pollInterval is used by the pollUntilTrue function to define the frequency that
+	// the provided function is ran
```
Comment (aravindhp) [inline test/e2e/csc_tests.go]: Not sure what you are trying to say. `function is run at`? Moreover did we make these constants so that any function that wants to poll can use it? If yes, please make the comment generic. Or is it specific to `pollUntilTrue()`? If yes, then you should move them into that function.

File: test/e2e/csc_tests.go [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+	pollInterval time.Duration = 5 * time.Second
+
+	// pollTimeout is used by the pollUntilTrue function to define the timeout
+	pollTimeout time.Duration = 1 * time.Minute
```
Comment (aravindhp) [inline test/e2e/csc_tests.go]: Same comment here.

File: test/e2e/marketplace_test.go [hunk 20]
```diff
@@ -74,7 +72,6 @@ func MarketplaceCluster(t *testing.T) {
 
 	// get global framework variables
 	f := test.Global
-
```
Comment (aravindhp) [inline test/e2e/marketplace_test.go]: https://github.com/operator-framework/operator-marketplace/pull/145#discussion_r269143459 applies here too.

File: test/e2e/marketplace_test.go [hunk 21]
```diff
@@ -126,7 +123,6 @@ func defaultCreateTest(t *testing.T, f *test.Framework, ctx *test.TestCtx) error
 	if err != nil {
 		return err
 	}
-
```
Comment (aravindhp) [inline test/e2e/marketplace_test.go]: https://github.com/operator-framework/operator-marketplace/pull/145#discussion_r269143459 applies here too

File: test/e2e/helpers.go (hunk 14 above)
Comment (awgreene) [reply test/e2e/helpers.go]: I have found a [similar method](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go#L305 ) in k8s apimachinery and will subsitute my function for it.

File: test/e2e/helpers.go [hunk 22]
```diff
@@ -71,3 +72,37 @@ func WaitForSuccessfulDeployment(t *testing.T, f *test.Framework, deployment app
 	t.Logf("Deployment %s has been initialized successfully\n", deployment.Name)
 	return nil
 }
+
+// createRuntimeObject creates a runtime object using the test framework
```
[the remark this answers is not in the document]
Comment (aravindhp) [reply test/e2e/helpers.go]: This is already in a separate file called `helper.go`. How about a dedicated package called `helpers` rather than prefixing each function?

File: test/e2e/marketplace_test.go [hunk 23]
```diff
@@ -61,9 +57,11 @@ func TestMarketplace(t *testing.T) {
 	if err != nil {
 		t.Fatalf("failed to add catalogsource custom resource scheme to framework: %v", err)
 	}
+
 	// run subtests
 	t.Run("marketplace-group", func(t *testing.T) {
 		t.Run("Cluster", MarketplaceCluster)
+		t.Run("Cluster", runCSCWithInvalidTargetNamespace)
```
[the remark this answers is not in the document]
Comment (awgreene) [reply test/e2e/marketplace_test.go]: Sounds good to me - are you okay doing this in the [refactor PR](https://github.com/operator-framework/operator-marketplace/pull/145#discussion_r269083130) we had discussed?

File: test/e2e/helpers.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (awgreene) [reply test/e2e/helpers.go]: The helper.go file was generated by the sdk - not a big issue given that were appending new methods to it but something worth highlighting.

[the remark this answers is not in the document]
Comment (awgreene) [reply test/e2e/helpers.go]: If we do want to move this to a separate package could we do it in the refactor PR mentioned earlier?

**Comment by awgreene** [comment]
@aravindhp @kevinrizza @tkashem all feedback has been addressed.

File: test/e2e/helpers.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (kevinrizza) [reply test/e2e/helpers.go]: I don't believe helper.go was created by the sdk, I'm fairly sure I created that manually.

**Comment by awgreene** [comment]
Install failure:
>level=fatal msg="failed to fetch Cluster: failed to generate asset \"Cluster\": failed to create cluster: failed to apply using Terraform"

/retest

File: test/e2e/helpers.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (awgreene) [reply test/e2e/helpers.go]: @kevinrizza My mistake, I had confused the helper functions you wrote with [similar helper functions provided by the sdk](https://github.com/operator-framework/operator-sdk/blob/master/pkg/test/e2eutil/wait_util.go).

**Comment by aravindhp** [comment]
Removed my approval to give @tkashem a chance to review

**Comment by tkashem** [comment]
Talked to @awgreene, the feedback I had will be addressed in a separate PR.
/approve
